src/trace/dma-trace.c

```diff
@@ -126,17 +126,52 @@ static enum task_state trace_work(void *data)
  */
 int dma_trace_init_early(struct sof *sof)
 {
+	int ret;
+
+	/* If this assert is wrong then traces have been corrupting
+	 * random parts of memory. Some functions run before _and_ after
+	 * DMA trace initialization and we don't want to ask them to
+	 * never trace. So dma_trace_initialized() must be either
+	 * clearly false/NULL or clearly true, we can't tolerate random
+	 * uninitialized values in sof->dmat etc.
+	 */
+	assert(!dma_trace_initialized(sof->dmat));
+
 	sof->dmat = rzalloc(SOF_MEM_ZONE_SYS_SHARED, 0, SOF_MEM_CAPS_RAM, sizeof(*sof->dmat));
+	if (!sof->dmat) {
+		ret = -ENOMEM;
+		goto err;
+	}
```

Comment:
If a `SOF_MEM_ZONE_SYS_SHARED` allocation fails you'll get a panic, no need to check